handle bin strings, handle optional fields in "traverse" example

speleo3 opened the pull request:
- Accept **msgpack** bin type for strings (will emit warnings, but will not reject the input)
- handle optional fields in "traverse" example

speleo3 committed:
compile on CentOS 6

speleo3 committed:
suppress warning for msgpack-c version < 2

speleo3 committed:
MMTF_ACCEPT_MSGPACK_BIN_AS_STR

Accept msgpack bin type for strings

speleo3 committed:
handle optional fields in "traverse" example

speleo3 commented:
Travis build failed because of a change in the `msgpack-c` repo, see https://github.com/msgpack/msgpack-c/pull/572

gtauriello commented:
General question/concern (more related to the MMTF specs I guess): is it desired to allow strings to be stored as "Binary"?
In the C++ decoder prototype this would currently not work as any "Binary" object is treated as an MMTF binary to be decoded (which is how Binary objects are described in the MMTF specs). I could change that of course, but I am not sure if decoders in other languages could deal with that "softening" of the MMTF specs...

speleo3 commented:
@gtauriello thanks for raising the point. I agree that a strict spec is preferable. The reason why I submitted this pull request was, that when I created some modified MMTF files for testing using Python 2.7, it was really easy to unconsciously put binary strings into the msgpack container (this is different in Python 3, which has a strict type separation of bytes and str). I've implemented MMTF load support in PyMOL, and for a "consumer" application I'd rather be tolerant to input, instead of rejecting it. E.g. we don't expect an encoded array as a map key, so if you see bin type there, you can assume it was supposed to be a string object.

speleo3 commented:
@josemduarte @valasatava what are your thoughts on this?

josemduarte commented:
Apologies for late reply... In my opinion it would be better to be strict. I would support following the spec as close as possible and not deviate from it, especially as the project is still in its early development. Encoders should not encode strings as binaries and decoders should be strict about it. A good approach would be to refuse decoding such files but emit warnings anyway.

Ping @arose, what's your opinion?

arose commented:
I think we are currently the only ones creating encoding libraries so we should make sure there are no ambiguous files out there. Emitting a warning is good in any case!

speleo3 committed:
sync with pymol/contrib/mmtf-c

- fix MMTF_parser_fetch_float (regression of d9f11122)
- valgrind catches

https://sf.net/p/pymol/code/HEAD/tree/trunk/pymol/contrib/mmtf-c/

speleo3 committed:
comment out #define MMTF_ACCEPT_MSGPACK_BIN_AS_STR

Accepting bin as str is an inofficial relaxation of the MMTF spec

speleo3 committed:
Merge remote-tracking branch 'rcsb/master' into accept-bin-as-str
